### PR TITLE
[ADD] Add sale order lot selection

### PR DESCRIPTION
**Pull request by hurrinico**
This Module Allows you to select on sale order the production lot that will be delivered

### CODE REVIEW

File: sale_order_lot_selection/README.rst [hunk 1]
```diff
@@ -0,0 +1,29 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
+
+Sale Order Lot Selection
+========================
+
+This Module Allow you to select on sale order the production lot that will be delivered
```
Comment (pedrobaeza) [inline sale_order_lot_selection/README.rst]: s/Allow/allows

File: sale_order_lot_selection/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+{
+    'name': 'Sale Order Lot Selection',
+    'version': '1.0',
+    'category': 'Sales Management',
+    'author': "Agile Business Group",
+    'website': 'http://www.agilebg.com',
+    'license': 'AGPL-3',
+    'depends': ['sale', 'sale_stock', 'procurement', 'stock'],
+    'init_xml': [],
```
Comment (pedrobaeza) [inline sale_order_lot_selection/__openerp__.py]: Deprecated: remove it

File: sale_order_lot_selection/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    'data': ['view/sale_view.xml'],
+    'installable': True,
+    'active': True,
```
Comment (pedrobaeza) [inline sale_order_lot_selection/__openerp__.py]: Don't make it active by default.

File: sale_order_lot_selection/model/sale.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+class sale_order(models.Model):
+    _inherit = "sale.order"
+
+    def _prepare_order_line_procurement(
+            self, cr, uid, order, line, group_id=False, context=None):
+        res = super(
+            sale_order, self)._prepare_order_line_procurement(
+                cr, uid, order, line, group_id, context=context)
+        if line and line.lot_id:
+            res['lot_id'] = line.lot_id.id
```
Comment (pedrobaeza) [inline sale_order_lot_selection/model/sale.py]: Use new API and you'll avoid these conditions.

**Comment by pedrobaeza** [comment]
Please put an specific icon (and smaller) for the module, to not confuse with other functionality.

About functionality itself, I miss:
- Check that the selected lot has stock available.
- Reserve stock when confirming or saving the order.

File: sale_order_lot_selection/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>. #
+#                                                                       #
+#########################################################################
+
+
+{
+    'name': 'Sale Order Lot Selection',
+    'version': '1.0',
+    'category': 'Sales Management',
+    'author': "Agile Business Group",
```
Comment (tafaRU) [inline sale_order_lot_selection/__openerp__.py]: Please add OCA as author

**Comment by hurrinico** [comment]
Hi @pedrobaeza in my last commit i've made some changings.
Now i reserve lot on sale order confirm, i'm thinking on how to check lot's availability on sale order but it's quite difficult; you have some suggestion about it? 
Thank you in advance

**Comment by pedrobaeza** [comment]
Why don't you reserve instead quants (with a many2many) in the same manner that Odoo works? That way, you only need to lock that quants. By standard, you can't do that, but you can use same mechanism than this: https://github.com/odoomrp/odoomrp-wip/blob/8.0/mrp_lock_lot/models/stock_quant.py to forbid to use that quants.

**Comment by Viggor** [comment]
Have you done a similar thing for mrp like https://github.com/OCA/sale-workflow/pull/176 because we have done a similar feature for the sale https://github.com/OCA/sale-workflow/pull/175  and stock https://github.com/OCA/sale-workflow/pull/177.

**Comment by pedrobaeza** [comment]
@hurrinico, please check the PR against your branch and move this to merge the module.

**Comment by pedrobaeza** [comment]
Two flake8 issues:

```
./sale_order_lot_selection/model/stock.py:21:1: F401 'fields' imported but unused
./sale_order_lot_selection/model/sale.py:107:22: F821 undefined name 'SaleOrderLine'
```

**Comment by Viggor** [comment]
@pedrobaeza Flake8 issues fixed in the last two merge.

**Comment by eLBati** [comment]
@Viggor @hurrinico see travis https://travis-ci.org/OCA/sale-workflow/jobs/70190209
`Can't reserve products for lot 0000010`

**Comment by hurrinico** [comment]
@pedrobaeza @eLBati @Viggor i've made some changes to tests and now travis turns green

File: sale_order_lot_selection/model/procurement.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+# You should have received a copy of the                                #
+# GNU Affero General Public License                                     #
+# along with this program.  If not, see <http://www.gnu.org/licenses/>. #
+#                                                                       #
+#########################################################################
+
+from openerp import fields, models, api
+
+
+class procurement_order(models.Model):
```
Comment (pedrobaeza) [inline sale_order_lot_selection/model/procurement.py:25]: Please use CamelCase for class names

File: sale_order_lot_selection/model/sale.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>. #
+#                                                                       #
+#########################################################################
+
+from openerp import fields, models, api, _
+from openerp.exceptions import Warning
+from openerp.tools import float_compare
+
+
+class sale_order_line(models.Model):
```
Comment (pedrobaeza) [inline sale_order_lot_selection/model/sale.py]: Please use CamelCase for class names

File: sale_order_lot_selection/model/sale.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                 max(0, product_obj.qty_available),
+                 product_obj.uom_id.name)
+            warning = {'title': _('Configuration Error!'),
+                       'message': _(
+                           "Not enough stock ! : ") + warn_msg + "\n\n"}
+            res.update({'warning': warning})
+            return res
+
+
+class sale_order(models.Model):
```
Comment (pedrobaeza) [inline sale_order_lot_selection/model/sale.py]: Please use CamelCase for class names

File: sale_order_lot_selection/model/sale.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+from openerp import fields, models, api, _
+from openerp.exceptions import Warning
+from openerp.tools import float_compare
+
+
+class sale_order_line(models.Model):
+    _inherit = 'sale.order.line'
+
+    lot_id = fields.Many2one('stock.production.lot', 'Lot', copy=False)
```
Comment (pedrobaeza) [inline sale_order_lot_selection/model/sale.py]: Shouldn't this be restricted to the lots of the product? (a domain)

**Comment by pedrobaeza** [comment]
I also wonder if we should restrict the lot selection to the current available lots. If not, we should note in the README that the lot selection doesn't correspond with available lots in stock (it's only a priority), but I'm not sure that in this case, the module will be useful.

**Comment by eLBati** [comment]
Afaik, the module prevents to use unavailable lots.
I suggest to add a test for the failing assignment

**Comment by pedrobaeza** [comment]
Indeed there's a warning programmed in an onchange, but it has 2 problems:
- The warning is not preventing to select that lot.
- You have to click on each lot to see which one is available, so it can be very tedious and not valid for a product with lots of serial numbers.

I suggest you to pre-calculate in a many2many field the available lots using quants, calling the method for getting the preferred domain and mapping the lots of these quants.

**Comment by hurrinico** [comment]
@pedrobaeza I've added a test case in order to dimostrate that if a lot on a sale order is already restricted, you can't confirm the order.

**Comment by pedrobaeza** [comment]
Yeah, I know, but my remark is about the usability. Can you read it again?

**Comment by hurrinico** [comment]
Ok @pedrobaeza we have understand the problem, and we agree with you.
Have you some idea about the fix? (this link doesn't work anymore https://github.com/odoomrp/odoomrp-wip/blob/8.0/mrp_lock_lot/models/stock_quant.py ) 
I was thinking about a new module that add a computed fields (boolean) to identify an available's lot?

**Comment by pedrobaeza** [comment]
The idea is to have a many2many computed field available lots in the sale.order.line, that depends on the warehouse, product and the quantity. This field is calculated calling the self.env['stock.quant'].quants_get_prefered_domain with the corresponding arguments, and applying that domain to make a search. Then, make a mapped function to retrieve the lots associated (this doesn't remove duplicates).

**Comment by hurrinico** [comment]
Hi @pedrobaeza , we have find this solution; we have replaced the onchange on lot_id with an override to the method "product_id_change_with_wh" where we have called (as you suggest) quants_get_prefered_domain and extended the domain. Could be this a solution?

**Comment by hurrinico** [comment]
@eLBati you are right!! i have made the requested changes.

**Comment by hurrinico** [comment]
@eLBati i've fixed the indentation

**Comment by eLBati** [comment]
Tried on runbot:
- install module
- create new SO
- add a new line, selecting 'Ice Cream' product
- go to lot field and click on arrow

get

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 336, in old_api
    result = method(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/models.py", line 1725, in name_search
    return self._name_search(name, args, operator, limit=limit)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 463, in new_api
    result = method(self._model, cr, uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/models.py", line 1737, in _name_search
    ids = self._search(cr, user, args, limit=limit, context=context, access_rights_uid=access_rights_uid)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/models.py", line 4654, in _search
    query = self._where_calc(cr, user, args, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/models.py", line 4465, in _where_calc
    e = expression.expression(cr, user, domain, self, context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/osv/expression.py", line 646, in __init__
    self.parse(cr, uid, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/osv/expression.py", line 749, in parse
    self.stack = [ExtendedLeaf(leaf, self.root_model) for leaf in self.expression]
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/osv/expression.py", line 532, in __init__
    self.check_leaf(internal)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3113776-140-40501d/openerp/osv/expression.py", line 588, in check_leaf
    raise ValueError("Invalid leaf %s" % str(self.leaf))
ValueError: Invalid leaf id
```

**Comment by hurrinico** [comment]
@eLBati i've fixed with my last commit

**Comment by eLBati** [comment]
I'm closing this in favor of https://github.com/OCA/sale-workflow/pull/202 (I fixed last issues and squashed some commits)

Recent updates, which should fix the remaining problems, are
https://github.com/eLBati/sale-workflow/commit/57eb33260e20b53a05e6b1dc336fe192e5dcca34
https://github.com/eLBati/sale-workflow/commit/e671c1d5e31e31e23760d46663c5512ed94a32eb

**Pull request closed without merging**